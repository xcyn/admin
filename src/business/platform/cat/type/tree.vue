<template>
  <div>
    <ibps-tree
      :title="title"
      :width="width"
      :height="height"
      :data="treeData"
      :options="treeOptions"
      :contextmenus="hasContextmenu?treeContextmenus:[]"
      :position="position"
      :loading="loading"
      :identity="identity"
      @action-event="handleTreeAction"
      @node-click="handleNodeClick"
      @expand-collapse="handleExpandCollapse"
    />
    <!-- 分类编辑 -->
    <type-edit
      :id="editId"
      :parent-data="typeData"
      :is-private="isPrivate"
      :category-key="categoryKey"
      :visible="typeFormVisible"
      :title="editTitle"
      @callback="loadTreeData"
      @close="visible => typeFormVisible = visible"
    />
    <!-- 分类排序 -->
    <type-sort
      :id="editId"
      :visible="sortFormVisible"
      title="分类排序"
      @callback="loadTreeData"
      @close="visible => sortFormVisible = visible"
    />
  </div>
</template>

<script>
import { findTreeData, remove } from '@/api/platform/cat/type'
import ActionUtils from '@/utils/action'

import TypeEdit from '@/views/platform/cat/type/edit'
import TypeSort from '@/views/platform/cat/type/sort'

export default {
  components: {
    TypeEdit,
    TypeSort
  },
  props: {
    title: {
      type: String
    },
    identity: {
      type: String
    },
    categoryKey: {
      type: String,
      required: true
    },
    hasContextmenu: { // 是否有右键菜单
      type: Boolean,
      default: false
    },
    width: {
      type: [String, Number],
      default: 200
    },
    height: {
      type: [String, Number],
      default: 500
    },
    position: {
      type: String,
      default: 'west'
    }
  },
  data() {
    return {
      loading: true,
      typeFormVisible: false,
      sortFormVisible: false,
      editId: '', // 编辑dialog需要使用
      editTitle: '编辑分类',

      // 树配置
      treeOptions: { 'rootPId': '-1' },
      treeContextmenus: [
        { icon: 'add', label: '添加分类', key: 'add' },
        { icon: 'edit', label: '编辑分类', key: 'edit', rights: ['node'] },
        { icon: 'delete', label: '删除分类', key: 'remove', rights: ['node'] },
        { divided: true },
        { icon: 'sort', label: '分类排序', key: 'sort' }
      ],
      treeData: [],

      isPrivate: false,
      typeData: {} // 记录分类信息
    }
  },
  created() {
    this.loadTreeData()
  },
  beforeDestroy() {
    this.treeOptions = null
    this.treeContextmenus = null
    this.treeData = null
    this.typeData = null
  },
  methods: {
    loadTreeData() {
      this.loading = true
      findTreeData({ 'categoryKey': this.categoryKey }).then(response => {
        this.loading = false
        this.treeData = response.data || []
      }).catch(() => {
        this.loading = false
      })
    },
    handleTreeAction(command, position, selection, data) {
      switch (command) {
        case 'refresh':// 查询
          this.loadTreeData()
          break
        case 'add':// 添加
          this.typeData = data // { 'name': data.name, 'id': data.id }
          this.isPrivate = false
          this.editTitle = '添加分类'
          this.handTreeEdit()
          break
        case 'addPrivate':// 添加
          this.typeData = data // { 'name': data.name, 'id': data.id }
          this.isPrivate = true
          this.editTitle = '添加私有分类'
          this.handTreeEdit()
          break
        case 'edit':// 编辑
          this.typeData = data
          this.isPrivate = true
          this.editTitle = '编辑分类'
          this.handTreeEdit(data.id)
          break
        case 'remove':// 删除
          this.handleTreeRemove(data.id)
          break
        case 'export':// 导出
          this.handTreeExport(data.id, data.name)
          break
        case 'sort':// 排序
          this.handTreeSort(data)
          break
      }
    },
    handleNodeClick(data) {
      this.$emit('node-click', data.parentId === '-1' ? '' : data.id)
    },
    handleExpandCollapse(isExpand) {
      this.$emit('expand-collapse', isExpand)
    },
    handTreeEdit(editId) {
      this.editId = editId || ''
      this.typeFormVisible = true
    },
    handTreeSort(data) {
      const children = data.children
      if (children && children.length > 0) {
        if (children.length === 1) {
          ActionUtils.warning('只有一个节点无需排序')
        } else {
          this.editId = data.id || ''
          this.sortFormVisible = true
        }
      } else {
        ActionUtils.warning('无子节点排序')
      }
    },
    handleTreeRemove(ids) {
      ActionUtils.removeRecord(ids).then((ids) => {
        remove({ typeId: ids }).then(response => {
          ActionUtils.removeSuccessMessage()
          this.loadTreeData()
        }).catch((err) => {
          console.error(err)
        })
      }).catch(() => { })
    }
  }
}
</script>
