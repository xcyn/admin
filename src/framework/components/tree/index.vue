<template>
	<div class="z-tree">
		<div class="tools">
			<el-button type="primary" icon="el-icon-refresh" @click="onRefresh" />
			<el-button type="primary" icon="fa fa-expand" @click="onToggle(true)" />
			<el-button type="primary" icon="fa fa-compress" @click="onToggle(false)" />
		</div>

		<div class="search">
			<el-input v-model="filterText" :placeholder="$t('common.field.enterKeywords')" />
			<el-button type="primary" icon="el-icon-search" @click="onSearch" />
		</div>

		<div class="tree">
			<el-tree ref="tree"
					 v-loading="loading"
					 v-bind="nativeProp"
					 :data="insideTreeData"
					 :node-key="nodeKey"
					 :show-checkbox="multipleSelect"
					 :props="propMapping"
					 :lazy="lazyLoad"
					 :highlight-current="true"
					 :filter-node-method="filterNode"
					 :load="loadNode"
           			 :default-expand-all="defaultAll"
					 @node-click="onNodeClick"
					 @check="onCheck"
					 @node-contextmenu="onRightClick" />
		</div>
	</div>
</template>

<script>
export default {
	props: {
		// 树组件数据
		treeData      : {
			type: [ Function, Array ],
			default(){
				return null
			},
		},
		// el-tree 原生属性
		nativeProp    : {
			type: Object,
			default(){
				return {}
			},
		},
		// key
		nodeKey       : {
			type: String,
			default(){
				return "id"
			},
		},
		// 是否允许多选
		multipleSelect: {
			type   : Boolean,
			default: true,
		},
    	//是否展开
    	defaultAll: {
			type   : Boolean,
			default: false,
		},
		// 属性映射
		propMapping   : {
			type: Object,
			default(){
				return {
					children: "children",
					label   : "label",
					filters : "",//根据这个字段过滤树节点,没有就使用label
				}
			},
		},
		// 是否懒加载
		lazyLoad      : {
			type   : Boolean,
			default: false,
		},
	},

	data(){
		return {
			filterText    : "",
			insideTreeData: [],
			insideloadNode: null,
			loading       : false,
		}
	},

	created(){
		const { treeData, lazyLoad } = this
		if(treeData instanceof Array){
			this.insideTreeData = [ ...treeData ]
		}else{
			this.loading        = true
			this.insideloadNode = treeData
			if(!lazyLoad){
				this.loadNode({ level: 0 }, this.setData)
			}
		}
	},

	methods: {
		/**
		 * 加载树节点
		 * @author mbb
		 */
		loadNode(node, resolve){
			if(node.level === 0){
				this._node    = node
				this._resolve = resolve
			}
			this.insideloadNode && this.insideloadNode(node, resolve, this.loadDone)
		},

		/**
		 * 重新加载整棵树
		 */
		onReload(){
			this.loadNode({ level: 0 }, this.setData)
		},

		/**
		 * 关闭loading加载
		 * @author mbb
		 */
		loadDone(){
			this.loading = false
		},

		/**
		 * 设置数据
		 */
		setData(treeData){
			this.insideTreeData = treeData
		},

		/**
		 * 树刷新
		 */
		onRefresh(){
			this.filterText = ""
			this.onSearch()
			this.$emit('tree-refresh')
		},

		/**
		 * 展开折叠
		 */
		onToggle(isExpand){
			let node = this.$refs.tree.store.currentNode
			//如果当前无节点被选中，默认设置第一个节点
			if(!node){
				node = this.$refs.tree.store.root
				//如果根节点无数据，设置第一个子节点
				if(!this.$refs.tree.store.root.label){
					node = this.$refs.tree.store.root.childNodes && this.$refs.tree.store.root.childNodes.length > 0 ? this.$refs.tree.store.root.childNodes[0] : ""
				}
			}
			if(!node){
				this.$message.warning("清先选择树节点")
			}else{
				node.expanded = isExpand
			}
			this.$forceUpdate()
		},

		/**
		 * 查询节点
		 * @author mbb
		 */
		onSearch(){
			this.$refs.tree.filter(this.filterText)
		},

		/**
		 * 过滤数据
		 */
		filterNode(value, data){
			if(!value){
				return true
			}
			if(this.propMapping.filters){
				return data[this.propMapping.filters].indexOf(value) !== -1
			}else{
				return data[this.propMapping.label].indexOf(value) !== -1
			}

		},

		/**
		 * 清除当前所选
		 * @author mbb
		 */
		clearClick(){
			this.$refs.tree.setCurrentKey(null)
		},

    /**
     * 设置指定key节点选中
     * @author mbb
     */
    setNodeCurrentKey(key){
    	this.$refs.tree.setCurrentKey(key)
    },

		/**
		 * 点击树节点
		 */
		onNodeClick(data){
			!this.multipleSelect && this.$emit("nodeClick", data)
		},

		/**
		 * 右击树节点
		 */
		onRightClick(MouseEvent, object, Node, element){
			this.$emit("rightClick", MouseEvent, object, Node, element)
		},

		/**
		 * 树节点多选时
		 */
		onCheck(data, nodes){
			this.multipleSelect && this.$emit("nodeClick", nodes)
		},
	},
}
</script>

<style lang="scss" scoped>
.z-tree {
	background: #fff;
	width: 100%;

	.tools {
		padding: 5px;
		border-top: 2px solid rgb(245, 245, 245);
		border-bottom: 2px solid rgb(245, 245, 245);
	}

	.search {
		padding: 5px;
		display: flex;
		border-bottom: 2px solid rgb(245, 245, 245);
	}

	.tree {
		width: 100%;
	}
  ::v-deep .el-tree--highlight-current .el-tree-node.is-current>.el-tree-node__content {
      background-color: #A9D0F5;
      width: 100%;
  }
}
</style>
