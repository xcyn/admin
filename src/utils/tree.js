/**
 * 树操作工具类</br>
 * Array转换Tree
 * <pre>
 * 作者:hugh zhuang
 * 邮箱:3378340995@qq.com
 * 日期:2018-07-02-下午3:29:34
 * 版权:广州流辰信息技术有限公司
 * </pre>
 */
import { defaultsDeep } from 'lodash'
const defaultSetting = {
  isParent: 'isParent',
  childrenKey: 'children',
  nameKey: 'name',
  titleKey: 'title',
  idKey: 'id',
  pIdKey: 'parentId',
  levelKey: 'level',
  rootPId: null
}
const treeUtil = {
  /**
 * 转换成树形结构
 * @param {*} data 数据
 * @param {*} setting 设置
 */
  transformToTreeFormat: function(data, setting = {}) {
    const sNodes = JSON.parse(JSON.stringify(data))
    setting = defaultsDeep({}, setting, defaultSetting)
    const idKey = setting.idKey
    const pIdKey = setting.pIdKey
    const childrenKey = setting.childrenKey

    let i, j, l
    if (!idKey || idKey === '' || !sNodes) { return [] }
    if (sNodes instanceof Array) {
      const tree = []
      const tmpMap = []// res_id:object
      for (i = 0, l = sNodes.length; i < l; i++) {
        tmpMap[sNodes[i][idKey]] = sNodes[i]
      }

      for (j = 0, l = sNodes.length; j < l; j++) {
        if (tmpMap[sNodes[j][pIdKey]] && sNodes[j][idKey] !== sNodes[j][pIdKey]) {
          if (!tmpMap[sNodes[j][pIdKey]][childrenKey]) {
            tmpMap[sNodes[j][pIdKey]][childrenKey] = []
          }
          tmpMap[sNodes[j][pIdKey]][childrenKey].push(sNodes[j])
        } else {
          tree.push(sNodes[j])
        }
      }
      return tree
    } else {
      return [sNodes]
    }
  },

  /**
   * 转换成树形结构(里面携带了按钮ID)
   * @author 马冰冰
   * @param {*} data 数据
   * @param {*} setting 设置
   */
  transformToTreeFormat2: function (data, setting = {}) {
    const sNodes = JSON.parse(JSON.stringify(data))
    setting = defaultsDeep({}, setting, defaultSetting)
    const idKey = setting.idKey
    const pIdKey = setting.pIdKey
    const childrenKey = setting.childrenKey

    // 如果没有指定主键是哪个,则无法构建树
    if (!idKey || idKey === '' || !sNodes) { return [] }

    if (sNodes instanceof Array) {
      const tree = [] // 最终要返回的树状数据
      const tmpMap = []// resId:object

      sNodes.forEach(node => { // 将数据先按照id:对象的形式封装起来
        const resId = node[idKey]
        tmpMap[resId] = node
      })

      sNodes.forEach(node => {
        const id = node[idKey] // 当前节点id
        const pid = node[pIdKey] // 父节点id
        const pNode = tmpMap[pid] // 父节点

        if (pNode && id !== pid) { // 如果还有父节点,则吧自己加入到父节点的子节点中
          if (!pNode[childrenKey]) {
            pNode[childrenKey] = []
          }
          if (!node[childrenKey] && node['resOptypePoList']) { // 处理叶子节点中的按钮
            if (node['resOptypePoList'].length !== 0) { // 叶子节点中有按钮
              node['resOptypePoList'].forEach(resOptypePo => {
                resOptypePo['name'] = resOptypePo['opetrationTypePo']['opTypeName']
              })
              node[childrenKey] = node['resOptypePoList']
            }
          }
          pNode[childrenKey].push(node)
        } else { // 如果没有父节点,则吧自己装进树
          tree.push(node)
        }
      })
      return tree
    }
    else {
      return [sNodes]
    }
  },
  /**
 * 转换成树结构 并设置了层级
 * @param {*} data 数据
 */
  transformToTreeAndLevelFormat: function(data, setting = {}) {
    setting = defaultsDeep({}, setting, defaultSetting)
    const node = treeUtil.transformToTreeFormat(data, setting)
    // 设置层级
    for (let i = 0; i < node.length; i++) {
      treeUtil.setSonNodeLevel(null, node[i], setting)
    }
    return node
  },
  /**
   * 树形转换成数组结构
   *
   */
  transformToArrayFormat: function(data, setting = {}) {
    if (!data) return []
    function _do(_node) {
      r.push(_node)
      const children = treeUtil.nodeChildren(_node, setting)
      if (children) {
        r = r.concat(treeUtil.transformToArrayFormat(children, setting))
      }
    }
    const nodes = JSON.parse(JSON.stringify(data))
    setting = defaultsDeep({}, setting, defaultSetting)
    let r = []
    if (nodes instanceof Array) {
      for (let i = 0, l = nodes.length; i < l; i++) {
        const node = nodes[i]
        _do(node)
      }
    } else {
      _do(nodes)
    }
    return r
  },
  nodeChildren: function(node, setting, newChildren) {
    if (!node) {
      return null
    }
    const key = setting.childrenKey
    if (typeof newChildren !== 'undefined') {
      node[key] = newChildren
    }
    return node[key]
  },
  /**
   * 设置儿子节点等级
   * @param {*} parentNode
   * @param {*} node
   */
  setSonNodeLevel: function(parentNode, node, setting) {
    if (!node) return
    const childrenKey = setting.childrenKey
    const levelKey = setting.levelKey

    node[levelKey] = parentNode ? parentNode[levelKey] + 1 : 0
    if (!node[childrenKey]) { return }
    for (let i = 0, l = node[childrenKey].length; i < l; i++) {
      if (node[childrenKey][i]) { treeUtil.setSonNodeLevel(node, node[childrenKey][i], setting) }
    }
  }
}

export default treeUtil
