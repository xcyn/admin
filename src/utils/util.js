/**
 * 全局的工具类
 * <pre>
 * 作者:hugh zhuang
 * 邮箱:3378340995@qq.com
 * 日期:2018-07-02-下午3:29:34
 * 版权:广州流辰信息技术有限公司
 * </pre>
 *
 * 可以使用 this.$utils.xx
 *  如: 判断是否为空
 *   this.$utils.isEmpty()
 */

import log from './util.log.js'
import cookies from './util.cookies.js'
import filterXSS from 'xss'
const rtrim = /^[\s\uFEFF\xA0]+|[\s\uFEFF\xA0]+$/g

const util = {
  cookies,
  log,
  /**
   * 判断是否为空
   * @param {*} obj
   */
  isEmpty: function(obj, allowBlank = false) {
    if (util.isNull(obj)) return true
    if (util.isArray(obj)) return obj.length === 0
    if (util.isString(obj)) return (!(allowBlank || obj.length > 0))
    if (util.isObject(obj)) return util.isEmptyObject(obj)
    for (var key in obj) if (obj.hasOwnProperty(key)) return false
    return obj === undefined || (!allowBlank ? obj === '' : false)
  },
  /**
   * 判断是否不为空
   * @param {*} obj
   */
  isNotEmpty: function(obj, allowBlank = false) {
    return !util.isEmpty(obj, allowBlank)
  },
  /**
   * 判断是否为空对象
   * @param {*} obj
   */
  isEmptyObject: function(obj) {
    if (!obj) return true
    for (const name in obj) {
      return false
    }
    return true
  },
  /**
   * 判断是否为不空对象
   * @param {*} obj
   */
  isNotEmptyObject: function(obj) {
    return !util.isEmptyObject(obj)
  },
  /**
   * 是否是对象
   * @param {*} input
   */
  isObject: function(input) {
    return Object.prototype.toString.call(input) === '[object Object]'
  },
  /**
   * 是否是数组
   * @param {*} input
   */
  isArray: function(input) {
    return input instanceof Array || Object.prototype.toString.call(input) === '[object Array]'
  },
  isDate: function(input) {
    return input instanceof Date || Object.prototype.toString.call(input) === '[object Date]'
  },
  isNumber: function(input) {
    return input instanceof Number || Object.prototype.toString.call(input) === '[object Number]'
  },
  isString: function(input) {
    return input instanceof String || Object.prototype.toString.call(input) === '[object String]'
  },
  isBoolean: function(input) {
    return typeof input === 'boolean'
  },
  isFunction: function(input) {
    return typeof input === 'function'
  },
  isNull: function(input) {
    return input === undefined || input === null
  },
  /**
   *  判断是否是数字（包含字符串数字和数字类型）
   * @param {*} input
   */
  isNum: function(input) {
    if (util.isEmpty(input)) {
      return false
    }
    if (util.isString(input) && (input.substr(input.length - 1, 1) !== '.')) { // 避免最后一位为小数
      input = Number(input)
    }
    if (isNaN(input)) {
      return false
    } else {
      return util.isNumber(input)
    }
  },
  isValidNumber: function(t) {
    return typeof t === 'number' && !isNaN(t) && isFinite(t)
  },
  isPlainObject: function(obj) {
    if (obj && Object.prototype.toString.call(obj) === '[object Object]' && obj.constructor === Object && !hasOwnProperty.call(obj, 'constructor')) {
      var key
      for (var k in obj) {
        key = k
      }
      return key === undefined || hasOwnProperty.call(obj, key)
    }
    return false
  },
  isJSON: function(str) {
    if (util.isString(str)) {
      try {
        const obj = JSON.parse(str)
        return util.isPlainObject(obj) || util.isArray(obj)
      } catch (e) {
        return false
      }
    }
    return false
  },
  trim: function(text) {
    return text == null ? '' : (text + '').replace(rtrim, '')
  },
  /**
   * 判断参数是否是其中之一
   */
  oneOf: function(value, validList) {
    for (let i = 0; i < validList.length; i++) {
      if (value === validList[i]) {
        return true
      }
    }
    return false
  },
  /**
   * 判断参数是否是数组对象其中之一
   */
  oneOfObj: function(value, validList, key) {
    for (let i = 0; i < validList.length; i++) {
      if (value === validList[i][key]) {
        return true
      }
    }
    return false
  },
  /**
   * 全局唯一标识符
   */
  guid: function() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
      const r = Math.random() * 16 | 0
      const v = c === 'x' ? r : (r & 0x3 | 0x8)
      return v.toString(16)
    })
  },
  /**
   * 转boolean 值
   * @param {*} obj
   * @param {*} defaultValue
   */
  toBoolean: function(obj, defaultValue = false) {
    if (util.isEmpty(obj)) {
      return defaultValue
    }
    return util.oneOf(obj, [true, 'True', 'Yes', 'true', '1', 1, 'yes', 'Y', 'y', 'T', 't'])
  },
  /**
 * 创建新数据，避免对象引用
 * @param {*} data
 * @param {*} defaultValue
 */
  newData: function(data, defaultValue) {
    return util.isNotEmpty(data) ? JSON.parse(JSON.stringify(data)) : (defaultValue || data)
  },
  /**
 * 转换json字符串的转换
 * @param {*} data
 * @param {*} defaultValue
 */
  parseData: function(data, defaultValue) {
    if (util.isNotEmpty(data)) {
      // eslint-disable-next-line no-eval
      return util.isPlainObject(data) || util.isArray(data) ? data : window.eval('(' + data + ')')
    } else {
      return (defaultValue || data)
    }
  },
  /**
 * 转换json字符串的转换
 * @param {*} data
 * @param {*} defaultValue
 */
  parseJSON: function(data, defaultValue) {
    if (util.isNotEmpty(data)) {
      return util.isJSON(data) ? JSON.parse(data) : data
    } else {
      return (defaultValue || data)
    }
  },
  /**
  * eval 数据
  */
  evalData: function(data) {
    // eslint-disable-next-line no-eval
    return window.eval(data)
  },
  /**
   * 格式化文件大小, 输出成带单位的字符串
   * @method formatSize
   * @grammar util.formatSize( size ) => String
   * @grammar util.formatSize( size, pointLength ) => String
   * @grammar util.formatSize( size, pointLength, units ) => String
   * @param {Number} size 文件大小
   * @param {Number} [pointLength=2] 精确到的小数点数。
   * @param {Array} [units=[ 'B', 'K', 'M', 'G', 'TB' ]] 单位数组。从字节，到千字节，一直往上指定。如果单位数组里面只指定了到了K(千字节)，同时文件大小大于M, 此方法的输出将还是显示成多少K.
   * @example
   * console.log( util.formatSize( 100 ) );    // => 100B
   * console.log( util.formatSize( 1024 ) );    // => 1.00K
   * console.log( util.formatSize( 1024, 0 ) );    // => 1K
   * console.log( util.formatSize( 1024 * 1024 ) );    // => 1.00M
   * console.log( util.formatSize( 1024 * 1024 * 1024 ) );    // => 1.00G
   * console.log( util.formatSize( 1024 * 1024 * 1024, 0, ['B', 'KB', 'MB'] ) );    // => 1024MB
   */
  formatSize: function(size, pointLength, units) {
    units = units || ['B', 'K', 'M', 'G', 'TB']
    let unit
    while ((unit = units.shift()) && size > 1024) {
      size = size / 1024
    }
    return (unit === 'B' ? size : size.toFixed(pointLength || 2)) + unit
  },
  /**
   * 格式化文本
   * @param {*} text
   */
  formatText: function(text) {
    return text !== null ? ('' + filterXSS(text)).replace(/([^>\r\n]?)(\r\n|\n\r|\r|\n)/g, '$1' + '<br />' + '$2') : ''
  },
  /**
   * @description 打开新页面
   * @param {String} url 地址
   */
  open: function(url, target = '_blank') {
    const a = document.createElement('a')
    a.setAttribute('href', url)
    a.setAttribute('target', target)
    a.setAttribute('id', 'ibps-open-link')
    document.body.appendChild(a)
    a.click()
    document.body.removeChild(document.getElementById('ibps-open-link'))
  },
  /**
 * 将array递归为一维数组。
 * @param {*} ary
 * @param {*} predicate
 * @param {*} result
 */
  flatten: function(ary, predicate, result) {
    result = result || []
    if (ary) {
      for (let i = 0; i < ary.length; i++) {
        const value = ary[i]
        if (Array.isArray(value)) {
          util.flatten(value, predicate, result)
        } else {
          (predicate && !predicate(value)) || result.push(value)
        }
      }
    }
    return result
  },
  //判断是否是Base64
  isBase64:function(str){
      if(str === '' || str.trim() === ''){
          return false;
      }
      try{
          return btoa(atob(str)) == str;
      }catch(err){
          return false;
      }
    },
  requestAnimationFrame(callbak, time = 50) {
    let raf = window.requestAnimationFrame || window.webkitRequestAnimationFrame || window.mozRequestAnimationFrame
    if (!raf) {
      raf = (fn) => setTimeout(fn, time)
    }
    raf(() => {
      setTimeout(callbak, time)
    })
  },
  /**
   * 动态加载样式
   * @param {*} styleText
   */
  createStyles(styleText, id) {
    const style = document.createElement('style')
    style.setAttribute('type', 'text/css')
    if (id) {
      const styleTag = document.getElementById(id)
      if (styleTag) {
        styleTag.parentNode.removeChild(styleTag)
      }
      style.setAttribute('id', id)
    }
    if (style.styleSheet) {
      style.styleSheet.cssText = styleText
    } else {
      style.appendChild(document.createTextNode(styleText))
    }
    (document.querySelector('head') || document.body).appendChild(style)
    return style
  },
  /**
   * 通过format获取el-date-picker的type
   * @param {*} format
   */
  getTypeByFormat(format) {
    let result = 'date'
    const reqex = {
      year: /^yyyy$/,
      month: /^yyyy(-){0,1}(M){1,2}$/,
      date: /^yyyy(-){0,1}(M){1,2}(-){0,1}(d){1,2}$/,
      datetime: /^yyyy(-){0,1}(M){1,2}(-){0,1}(d){1,2} ((H){1,2}|(h){1,2}):(m){1,2}:(s){1,2}$/,
      time: /^((((H){1,2}|(h){1,2}):(m){1,2}:(s){1,2})|(((H){1,2}|(h){1,2}):(m){1,2})|((m){1,2}:(s){1,2}))$/
    }
    if (reqex.year.test(format)) {
      result = 'year'
    } else if (reqex.month.test(format)) {
      result = 'month'
    } else if (reqex.date.test(format)) {
      result = 'date'
    } else if (reqex.datetime.test(format)) {
      result = 'datetime'
    } else if (reqex.time.test(format)) {
      result = 'time'
    }
    return result
  },
  // 金额添加千分位
  comdify(n) {
    if (util.isEmpty(n)) return n
    const str = n.toString().split('.')
    const re = /\d{1,3}(?=(\d{3})+$)/g
    str[0] = str[0].replace(re, '$&,')
    return str.join('.')
  },
  // 去除千分位中的‘，’
  delcommafy(num) {
    if (util.isEmpty(num)) return num
    num = num.toString()
    num = num.replace(/,/gi, '')
    return num
  },
  // 对象比较
  valueEqual(obj1, obj2) {
    const o1 = obj1 instanceof Object
    const o2 = obj2 instanceof Object
    // 判断是不是对象
    if (!o1 || !o2) {
      return obj1 === obj2
    }

    if (Object.keys(obj1).length !== Object.keys(obj2).length) {
      return false
    }

    for (var o in obj1) {
      const t1 = obj1[o] instanceof Object
      const t2 = obj2[o] instanceof Object
      if (t1 && t2) {
        return this.valueEqual(obj1[o], obj2[o])
      } else if (obj1[o] !== obj2[o]) {
        return false
      }
    }
    return true
  }
}

export default util
