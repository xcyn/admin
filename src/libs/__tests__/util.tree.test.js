import { treeify } from '../util.tree';

// const tree = [
//   {
//     id: 1,
//     label: '一级 1',
//     value: '1',
//     children: [{
//       id: 11,
//       label: '二级 1-1',
//       value: '1-1',
//       children: [{
//         id: 111,
//         label: '三级 1-1',
//         value: '1-1-1'
//       }]
//     }]
//   },
//   {
//     id: 2,
//     label: '一级 2',
//     value: '2',
//     children: [{
//       id: 21,
//       label: '二级 2-1',
//       value: '2-1'
//     }, {
//       id: 22,
//       label: '二级 2-2',
//       value: '2-2'
//     }]
//   },
//   {
//     id: 3,
//     label: '一级 3',
//     value: '3',
//     children: [{
//       id: 31,
//       label: '二级 3-1',
//       value: '3-1'
//     }, {
//       id: 32,
//       label: '二级 3-2',
//       value: '3-2'
//     }]
//   }
// ];

const list = [
  {
    id: 1,
    label: '一级 1',
    value: '1',
    parentId: null
  },
  {
    id: 11,
    label: '二级 1-1',
    value: '1-1',
    parentId: 1
  },
  {
    id: 111,
    label: '三级 1-1',
    value: '1-1-1',
    parentId: 11
  },
  {
    id: 2,
    label: '一级 2',
    value: '2',
    parentId: null
  },
  {
    id: 21,
    label: '二级 2-1',
    value: '2-1',
    parentId: 2
  },
  {
    id: 22,
    label: '二级 2-2',
    value: '2-2',
    parentId: 2
  },
  {
    id: 3,
    label: '一级 3',
    value: '3',
    parentId: null
  },
  {
    id: 31,
    label: '二级 3-1',
    value: '3-1',
    parentId: 3
  },
  {
    id: 32,
    label: '二级 3-2',
    value: '3-2',
    parentId: 3
  }
];

describe('util.tree', () => {
  it('treeify', () => {
    expect(treeify(list, { id: 'id', parentId: 'parentId', children: 'children' })).toMatchSnapshot();
  });
});
