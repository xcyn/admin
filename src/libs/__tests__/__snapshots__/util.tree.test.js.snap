// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util.tree treeify 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "id": 111,
            "label": "三级 1-1",
            "parentId": 11,
            "value": "1-1-1",
          },
        ],
        "id": 11,
        "label": "二级 1-1",
        "parentId": 1,
        "value": "1-1",
      },
    ],
    "id": 1,
    "label": "一级 1",
    "parentId": null,
    "value": "1",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "id": 21,
        "label": "二级 2-1",
        "parentId": 2,
        "value": "2-1",
      },
      Object {
        "children": Array [],
        "id": 22,
        "label": "二级 2-2",
        "parentId": 2,
        "value": "2-2",
      },
    ],
    "id": 2,
    "label": "一级 2",
    "parentId": null,
    "value": "2",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "id": 31,
        "label": "二级 3-1",
        "parentId": 3,
        "value": "3-1",
      },
      Object {
        "children": Array [],
        "id": 32,
        "label": "二级 3-2",
        "parentId": 3,
        "value": "3-2",
      },
    ],
    "id": 3,
    "label": "一级 3",
    "parentId": null,
    "value": "3",
  },
]
`;
